pkg/controllers/updaterun/validation_test.go

```diff
@@ -0,0 +1,358 @@
+				Clusters: []placementv1alpha1.ClusterUpdatingStatus{
+					{ClusterName: "cluster-1"},
+				},
+			},
+			wantErr:                    wrapErr(true, fmt.Errorf("cluster `cluster-1` in the finished stage `test-stage` has not succeeded")),
+			wantUpdatingStageIndex:     -1,
+			wantLastFinishedStageIndex: -1,
+		},
+		{
+			name:                   "determineUpdatingStage should return error if the finished stage is not right after the last finished stage",
+			curStage:               2,
+			updatingStageIndex:     -1,
+			lastFinishedStageIndex: 0,
+			stageStatus: &placementv1alpha1.StageUpdatingStatus{
+				StageName:  "test-stage",
+				Conditions: []metav1.Condition{generateTrueCondition(updateRun, placementv1alpha1.StageUpdatingConditionSucceeded)},
+			},
+			wantErr:                    wrapErr(true, fmt.Errorf("the finished stage `test-stage` is not right after the last finished stage with index `0`")),
+			wantUpdatingStageIndex:     -1,
+			wantLastFinishedStageIndex: -1,
+		},
+		{
+			name:                   "determineUpdatingStage should return error if some stage has failed",
+			curStage:               0,
+			updatingStageIndex:     -1,
+			lastFinishedStageIndex: -1,
+			stageStatus: &placementv1alpha1.StageUpdatingStatus{
+				StageName:  "test-stage",
+				Conditions: []metav1.Condition{generateFalseCondition(updateRun, placementv1alpha1.StageUpdatingConditionSucceeded)},
+			},
+			wantErr:                    wrapErr(false, fmt.Errorf("the stage `test-stage` has failed, err: ")),
+			wantUpdatingStageIndex:     -1,
+			wantLastFinishedStageIndex: -1,
+		},
+		{
+			name:                   "determineUpdatingStage should return error if there are multiple stages updating",
+			curStage:               1,
+			updatingStageIndex:     0,
+			lastFinishedStageIndex: -1,
+			stageStatus: &placementv1alpha1.StageUpdatingStatus{
+				StageName:  "test-stage",
+				Conditions: []metav1.Condition{generateTrueCondition(updateRun, placementv1alpha1.StageUpdatingConditionProgressing)},
+			},
+			wantErr:                    wrapErr(true, fmt.Errorf("the stage `test-stage` is updating, but there is already a stage with index `0` updating")),
+			wantUpdatingStageIndex:     -1,
+			wantLastFinishedStageIndex: -1,
+		},
+		{
+			name:                   "determineUpdatignStage should return error if the updating stage is not right after the last finished stage",
+			curStage:               1,
+			updatingStageIndex:     -1,
+			lastFinishedStageIndex: -1,
+			stageStatus: &placementv1alpha1.StageUpdatingStatus{
+				StageName:  "test-stage",
+				Conditions: []metav1.Condition{generateTrueCondition(updateRun, placementv1alpha1.StageUpdatingConditionProgressing)},
+			},
+			wantErr:                    wrapErr(true, fmt.Errorf("the updating stage `test-stage` is not right after the last finished stage with index `-1`")),
+			wantUpdatingStageIndex:     -1,
+			wantLastFinishedStageIndex: -1,
+		},
+		{
```

Comment:
Do we have the case
 stage update status cannot be true if there are still cluster updating in it